Add a HowTo on searching the Hub efficiently (#576)
huggingface/huggingface_hub: docs/hub/searching-the-hub.md

# Searching the Hub Efficiently with Python

In this tutorial we will explore how to interact and search the Hub with the `HfApi` class to help find available models and datasets quickly.

## The Basics

> The `huggingface_hub` comes packaged with an interface that can interact with the Hub in the `HfApi` class:

Author: Went with:
`huggingface_hub` is a Python library that allows anyone to freely extract useful information from the Hub, as well as downloading and publishing models. You can install it with:
```bash
pip install huggingface_hub
```

It comes packaged with an interface that can interact with the Hub in the `HfApi` class:

Our first criteria is we want all PyTorch models. This would be found under the `library` attribute, so let's see if it is there:

```python
>>> model_args.library
```

    Available Attributes or Keys:
     * AdapterTransformers
     * Asteroid
     * ESPnet
     * Flair
     * JAX
     * Joblib
     * Keras
     * ONNX
     * PyTorch
     * Pyannote
     * Rust
     * Scikit_learn
     * SentenceTransformers
     * Stanza
     * TFLite
     * TensorBoard
     * TensorFlow
     * TensorFlowTTS
     * Timm
     * Transformers
     * allennlp
     * fastText
     * fastai
     * spaCy
     * speechbrain

We can see that it is under `PyTorch`. So what we want to use is `model_args.library.PyTorch`:

```python
>>> model_args.library.PyTorch
```

    'pytorch'

Below is a graphic repeating the process for finding both our `Text Classification` requirement, as well as our `glue` requirement:

![](../assets/hub/search_text_classification.gif)

![](../assets/hub/search_glue.gif)

> Now that we have the pieces, we just need to combine them all together for something the api can use, through the `ModelFilter` and `DatasetFilter` classes. These take those outputs we just had, and transform them into something the API can use conviently:

Reviewer: ```suggestion
Now that all the pieces are there, the last step is to combine them all for something the API can use through the `ModelFilter` and `DatasetFilter` classes. The classes transform the outputs of the previous step into something the API can use conveniently:
```

```python

> >>> from huggingface_hub import ModelFilter, DatasetFilter

Reviewer: ```suggestion
>>> from huggingface_hub import ModelFilter
```

Removed as it was not used

Author: Opting to keep it in for a Dataset example at the end

    	sha: f89a85cb8703676115912fffa55842f23eb981ab
    	lastModified: 2021-12-21T14:03:01.000Z
    	tags: ['pytorch', 'tensorboard', 'distilbert', 'text-classification', 'dataset:glue', 'transformers', 'license:apache-2.0', 'generated_from_trainer', 'model-index', 'infinity_compatible']
    	pipeline_tag: text-classification
    	siblings: [ModelFile(rfilename='.gitattributes'), ModelFile(rfilename='.gitignore'), ModelFile(rfilename='README.md'), ModelFile(rfilename='config.json'), ModelFile(rfilename='pytorch_model.bin'), ModelFile(rfilename='special_tokens_map.json'), ModelFile(rfilename='tokenizer.json'), ModelFile(rfilename='tokenizer_config.json'), ModelFile(rfilename='training_args.bin'), ModelFile(rfilename='vocab.txt'), ModelFile(rfilename='runs/Dec21_13-51-40_bc62d5d57d92/events.out.tfevents.1640094759.bc62d5d57d92.77.0'), ModelFile(rfilename='runs/Dec21_13-51-40_bc62d5d57d92/events.out.tfevents.1640095117.bc62d5d57d92.77.2'), ModelFile(rfilename='runs/Dec21_13-51-40_bc62d5d57d92/1640094759.4067502/events.out.tfevents.1640094759.bc62d5d57d92.77.1')]
    	config: None
    	private: False
    	downloads: 6
    	library_name: transformers
    	likes: 0
    }

> As we can see, it found the models we wanted that fit all of our criteria. We can even take it further by passing in an array for each of the parameters we had before. For example, now we'll look for the same configuration, but also include `TensorFlow` in the filter:

Reviewer: ```suggestion
As you can see, it found the models that fit all the criteria. You can even take it further by passing in an array for each of the parameters from before. For example, let's take a look for the same configuration, but also include `TensorFlow` in the filter:
```
